knitr/convert-odes/convert_odes.Rmd

```diff
@@ -0,0 +1,398 @@
+function, and $y_0$ and $t_0$ are the initial conditions.
+
+The solution, $y(t, \theta)$, is written explicitly in terms of $\theta$
+as a reminder that the solution of an ODE initial value problem can be a
+function of model parameters or data. This is the usual use case: an ODE
+is parameterized by a set of parameters that are to be estimated.
+
+Specifying an ODE initial value problem in Stan involves writing a Stan
+function for the ODE system function, $f(t, y, \theta)$. The big difference
+in the old and new ODE interfaces is that previously any arguments meant for
+the ODE system function had to be manually packed and unpacked from special
+argument arrays that were passed from the ODE solve function call to the ODE
+system function -- in the new interface none of this packing and unpacking
+is necessary.
+
+The new interface also uses `vector` variables for the state rather than
+`real[]` variables.
+
+As an example, in the old solver interface the system function for the ODE
+$y' = -\alpha y$ would be written:
+
+```{stan, output.var = "", eval = FALSE}
+functions {
+  real[] rhs(real t, real[] y, real[] theta, real[] x_r, int[] x_i) {
+    real alpha = theta[1];
+
+    real yp[1] = { -alpha * y[1] };
+
+    return yp;
+  }
+}
+```
+
+In the new interface the system function can be written:
+
+```{stan, output.var = "", eval = FALSE}
+functions {
+  vector rhs(real t, vector y, real alpha) {
+    vector[1] yp = -alpha * y;
+
+    return yp;
+  }
+}
+```
+
+The new interface avoids unused arguments `x_r`, and `x_i`, and the
+parameter `alpha` can be passed directly instead of being packed into `theta`.
+
+For a simple function, this does not look like much, but for more
+complicated models with numerous arguments of different types, the packing
+and unpacking is tedious and error prone. This leads to models that are
+difficult to debug and difficult to iterate on.
+
+# New Interface
+
+The new interface introduces six new functions:
+
+`ode_bdf`, `ode_adams`, `ode_rk45` and `ode_bdf_tol`,`ode_adams_tol`,
+`ode_rk45_tol`
+
+The solvers in the first columns have default tolerance settings. The solvers
+in the second column accept arguments for relative tolerance, absolute
+tolerance, and the maximum number of steps to take between output times.
+
+This is different from the old interface where tolerances are presented
+as overloads to the same function.
+
+The new `ode_bdf` solver interface is (the interfaces for `ode_adams` and
+`ode_rk45` are the same):
+
+```{stan, output.var = "", eval = FALSE}
+vector[] ode_bdf(F f,
+                 vector y0,
+                 real t0, real[] times,
+                 T1 arg1, T2 arg2, ...)
+```
+
+The arguments are:
+
+1. ```f``` - ODE system function
+
+2. ```y0``` - Initial state of the ODE
+
+3. ```t0``` - Initial time of the ODE
+
+4. ```times``` - Sorted array of times to which the ode will be solved (each
+  element must be greater than t0)
```

Comment:
maybe emphasise that we now allow successive times to be the same. So maybe say

"Sorted (and not strictly sorted) array..."

Reply:
Although... the old interface now allows that as well... whatever...